### ISSUE TITLE
Why isn't the reference model re-initialized for each epoch in GRPO?

### ISSUE THREAD

**Issue by Jerrrrykun** [allenai/open-instruct#546]
Hi. Thanks for your efficient updates on the GRPO implementation!

When comparing the `grpo_vllm_thread_ray_gtrl.py` and the pseudocode from the Deepseek-Math paper, I found that the implemented GRPO trainer does not re-initialize the reference model for each epoch, which contradicts the pseudocode in the paper. (This also holds for PPOTrainer in this repo.)

<img width="938" alt="Image" src="https://github.com/user-attachments/assets/b928dd9c-c666-4bbe-86f2-8a32e9d4f290" />

https://github.com/allenai/open-instruct/blob/4afebdeda19e04d247174ecc90554948a4b3bbb0/open_instruct/grpo_vllm_thread_ray_gtrl.py#L1177-L1180

So, I wonder whether you also follow the engineering trend like this: https://github.com/huggingface/trl/issues/1112. (Did anyone try adding the reference model update step and demonstrate that in RL4LLM this could be skipped? I am just curious as the reinitialization does not need to take too much computation. Why not follow the pseudocodes?)

**Comment by vwxyzjn** [comment]
This is kind of interesting: also if the reference model is updated very frequently, would it be the same as not having a reference model at all?

**Comment by hamishivi** [comment]
It would require a bit more changes. The snippet you linked to is actually the `ppo_epochs`, which are different to the data-level epochs (confusing, sorry) -- this is because in RL sometimes people sample data, and then perform a few passes (epochs) of training on it with the PPO objective. Actually, by default we do 4 passes over it in this repo!

To properly change, we would need to track when the training_step passes into a new epoch, and update the reference model accordingly.

**Comment by hamishivi** [comment]
Also: technically, yes, once you take more than 1 step we are off-policy. In the deepseek paper, they say they do 1 'ppo epoch' (i.e., remain on-policy). But in that case, all the clipping logic used in PPO and included in the GRPO paper doesn't come into effect. PPO/GRPO includes this to ensure we don't go 'too far' from the original policy, so our updates are still 'on-policy-ish'

@vwxyzjn feel free to correct me if I'm slightly wrong here 😅

**Comment by hamishivi** [comment]
For this, we actually do resample at each step! 

To break it down, at each train step:
1. We sample from the model at lines 999-10243, using vllm (https://github.com/allenai/open-instruct/blob/main/open_instruct/grpo_vllm_thread_ray_gtrl.py#L999-L1043)
2. We post-process the responses, compute reward scores, etc. lines 1055-1194. (https://github.com/allenai/open-instruct/blob/main/open_instruct/grpo_vllm_thread_ray_gtrl.py#L1055-L1194)
3. We take training steps over these responses, potentially multiple passes, and potentially in multiple batches in lines 1198-1291. (https://github.com/allenai/open-instruct/blob/main/open_instruct/grpo_vllm_thread_ray_gtrl.py#L1198-L1291). You can setup the hyperparameters such that you only do one pass and one batch (or simulate one batch with gradient accumulation).
4. We repeat the process. We call steps 1-4 a grpo/ppo training step.

So our samples are on-policy, but you can configure to take more and more steps on the responses, which becomes more 'offline'. the PPO clipping logic will help here, but yes, performance will drop if you go more 'offline', usually.

One final caveat is that we usually run our code *aync*, which means we overlap generation and training -- we let the model take a train step while we generate samples from the previous step. So, technically, the samples are usually from the model 1 step prior. There is some past work showing this doesn't harm performance (https://arxiv.org/abs/2410.18252), potentially because only being 1 step off isn't that bad (vs being many steps off).

Hopefully that makes sense! The code is definitely quite complicated.
